### PR TITLE
Added "append_array" function to address Issue 481

### PR DESCRIPTION
**Pull request by bbbales2**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Addresses: https://github.com/stan-dev/math/issues/481

I called it append_array instead of append to emphasize that it's for arrays only (like rep_array, as compared to max/min/sum, etc. that work with both).

I have a few questions.

1. Should this work for scalars of type T?

```
T[] append(T[] a, T b);
T[] append(T a, T[] b);
```

2. Should this be called concatenate_array instead?

3. Should this work for vectors/matrices or is that best left untouched (seems like it could get a lot more involved here)?

4. Seems trivial, but I want to be clear, this sort of copying around vars isn't going to add any complexity to the autodiff tree, right?

(I'll head off and find a more substantial issue next -- don't mean to take all the easy ones :p)

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): University of California, Santa Barbara

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: stan/math/prim/arr/fun/append_array.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    /**
+     * Concatenate two vectors into one.
+     *
+     * @tparam T Type of elements contained in vector.
+     * @param x First vector.
+     * @param y Second vector.
+     * @return A vector of x and y concatenated together (in that order).
+     */
+    template <typename T>
+    inline std::vector<T> append_array(std::vector<T> x, std::vector<T> y) {
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/append_array.hpp]: The arguments need to be declared as `const std::vector<T>&` --- a reference to a constant.

File: stan/math/prim/arr/fun/append_array.hpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+#define STAN_MATH_PRIM_ARR_FUN_APPEND_ARRAY_HPP
+
+#include <vector>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * Concatenate two vectors into one.
+     *
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/append_array.hpp]: You want to write the first sentence of doc to describe what it does either in terms of return or side effect if there's not a return.  This should be something like "Return the concatenation of the two specified vectors in the order of the arguments."

File: test/unit/math/prim/arr/fun/append_test.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#include <stan/math/prim/arr.hpp>
+#include <gtest/gtest.h>
+
+TEST(MathFunctions, append_array) {
```
Comment (bob-carpenter) [inline test/unit/math/prim/arr/fun/append_test.cpp]: You generally want to test boundary conditions, like one or both of the vectors being size zero.  I know the rest of our tests haven't been that thorough.

**Comment by bob-carpenter** [comment]
You probably want to talk to us before making these requests.  There are some more things that will have to be done if this is going to fly in Stan.  In the Stan language, any variables can be put into any arguments.  That means that you won't always have heterogeneous appends.  You might do this:

```
data {
  real a[10];
parameters {
  real b[20];
transformed parameters }
  real ab = append(a, b);
```

So that's going to require at least these three signatures:

```
template <typename T>
vector<T> append(const vector<T>& a, const vector<double>& b);

template <typename T>
vector<T> append(const vector<double>& a, const vector<T>& b);

vector<T> append(const vector<T>& a, const vector<T>& b);
```

So what you probably want to do is something like this:

```
template <typename T1, typename T2>
typename boost::promote_args<T1, T2>::type
append(const vector<T1>& a, const vector<T2>& b) {
  vector<typename boost::promote_args<T1, T2>::type> ab;
  ... and then the rest of your code should work because a and b will be assignable ...
}
```

This alas won't deal with the integer case(s).  If you want to be really generous, you'd let integer arrays be concatenated by promoting to double arrays, but that's not necessary---we don't have that full type promotion elsewhere.

**Comment by bbbales2** [comment]
>> You probably want to talk to us before making these requests.

Good point -- seems obvious in retrospect. I'll send a note before I attack another issue so I don't wander too far off on my own.

Changes make sense. I'll get em' plugged in and uploaded soon as I get a chance!

**Comment by bbbales2** [comment]
I made the changes. Is the naming okay? (append_array instead of append)? Seems consistent with the other append_*s.

**Comment by bob-carpenter** [comment]
Yes, naming is fine.

**Comment by bbbales2** [comment]
Take 3!

**Comment by bob-carpenter** [comment]
Easier to write as a comment than another review.  

First off, thanks for testing the derivatives.  We might have skipped that here because we know this function doesn't do anything and we don't have a good general derivative testing framework yet (I'm working on that now).

Second, we ideally want to test all possible instantiations of functions.  In `rev`, the one we're using now and most care about, that's

```
vector<var>, vector<double>
vector<double>, vector<var>
vector<var>, vector<var>
```

We know these will be OK with the current implementation, but partly what we try to do with the tests is make sure nobody breaks the functions in the future.  We've been burned many times before with functions not being fully instantiatable. This will get tested again for actually being instantiated in Stan with a sample program, so it may not be such a big deal here.

Ideally we would do the same with `fvar<double>` and `fvar<fvar<double> >` in place of `var` in forward, and then `fvar<var>`, `fvar<fvar<var> >` in mixed.  That's up to third order derivatives and all we anticipate using in the near term (it's part of RHMC).  We don't actually use the `fvar<double>` or `fvar<fvar<double> >` anywhere other than in the math lib for completeness.  But you really don't need to worry about derivatives at this level.

I'm sorry there isn't a better derivative/value testing framework here.  It's a huge pain to explore all the combinatorics in tests like this.

Maybe others will have opinions about whether these extensive tests should be necessary for simple functions like this.  They're very incomplete now, but that's no excuse for not doing it right as we add new functions.  I'm also going to go back and fix all the tests for fwd and mix as my next big task.

**Comment by bbbales2** [comment]
Oh, yeah, I saw the fvar<fvar<...>> stuff and wasn't super sure what it was. That makes sense. Easy enough to add more tests in. Take 4!

**Comment by bob-carpenter** [comment]
OK.  I don't want to make submitting new functions too onerous.  This isn't a never-ending process and partly I just want to lay out what we're expecting.  It's a lot of tests, but they've saved us again and again from making dumb mistakes.

**Comment by bbbales2** [comment]
Take 4 is up!

In rev I included tests for:
vector<var>, vector<double>
vector<double>, vector<var>
vector<var>, vector<var>

In fwd I included test for:
vector<fvar>, vector<double>
vector<double>, vector<fvar>
vector<double>, vector<fvar<fvar>>
vector<fvar<fvar>>, vector<double>
vector<fvar>, vector<fvar>
vector<fvar<fvar>>, vector<fvar>
vector<fvar>, vector<fvar<fvar>>
vector<fvar<fvar>>, vector<fvar<fvar>>

Which in retrospect seems silly since these guys aren't really used but whatever.

In mix I included tests for:
vector<fvar<fvar<var>>>, vector<fvar<var>>
vector<fvar<var>>, vector<fvar<fvar<var>>>
vector<fvar<fvar<var>>>, vector<fvar<fvar<var>>>

**Comment by bbbales2** [comment]
Hmm, by not worry about the derivatives did you mean I just needed to check the .val().val() stuff? Not setting .d_s and checking they show up in the right place? I'm fine removing those derivative things if so. These tests are awkwardly complex for what they are.

**Comment by bob-carpenter** [comment]
Might as well leave them in now.  I'm going to try to go through and refactor all of our derivative testing
to use a general framework.  The new C++11 lambdas and bind are going to make this a lot easier.

**Comment by bbbales2** [comment]
Travis failed. I checked with g++ on my computer and it fails as well. I started this branch before the ambiguous fvar fix came along (https://github.com/stan-dev/math/issues/528), which seems to be the error I'm getting.

How about I rebase this branch on the new stuff in develop? I can also merge all my little commits into one. I'm pretty sure that'll kill this branch and I'll need to open a new pull request, but the only thing lost will be the comments here.

**Comment by bbbales2** [comment]
Yeah, something automated would be nice. I dunno what it'd look like, but even all that coding I did last night wasn't strictly speaking a complete set of tests.

**Comment by bbbales2** [comment]
Ooops! Looks like my fault. I made a copy of the code and rebased it on the new develop. Looks like I was abusing assumptions of the old fvar stuff, cause now the code now fails in g++ and clang++ with the same errors. I'll investigate. I'll send another note when I figure out what's happening.

**Comment by bob-carpenter** [comment]
I have a pending pull request on stan-dev/math that should fix the
combination of Eigen 3.3.3 and fvar problems.

- Bob

**Comment by bob-carpenter** [comment]
Let us know when this is ready to retest.  If you just need to kick off tests again, you can do this by asking Jenkins.

**Comment by bbbales2** [comment]
Moving this to: https://github.com/stan-dev/math/pull/550

**Pull request closed without merging**